File: src/graph/Carousel.js
import FanPainter from './FanPainter';
import {matrixMultiply3x3, makeScale3x3, makeTranslation3x3} from '../gl';
import {CAROUSEL_SHOW_DURATION} from './settings';
/* eslint-disable require-jsdoc */

export default function Carousel(viewport) {
  this._viewport = viewport;

  this._updateRepeatedly = false;
  this._showScale = 1;

  this.onScheduleRepaint = null;
  this.onScheduleRepaintThisArg = null;

  // Carousel-rendered carets.
  this._carouselPaintingDirty = true;
  this._carouselPlots = [];
  this._carouselCallbacks = [];

  // Location of the carousel, in world coordinates.
  this._carouselCoords = [0, 0];
  this._carouselSize = 50;

  this._showCarousel = false;
  this._selectedCarouselPlot = null;
  this._selectedCarouselPlotIndex = null;

  // GL painters are not created until needed.
  this._fanPainter = null;

  this._selectedPlot = null;
}

Carousel.prototype.window = function() {
  return this._viewport.window();
};

Carousel.prototype.camera = function() {
  return this._viewport.camera();
};

Carousel.prototype.needsRepaint = function() {
  return this._carouselPaintingDirty || this._updateRepeatedly;
};

Carousel.prototype.moveCarousel = function(worldX, worldY) {
  this._carouselCoords[0] = worldX;
  this._carouselCoords[1] = worldY;
};

Carousel.prototype.setCarouselSize = function(size) {
  this._carouselSize = size;
};

Carousel.prototype.showCarousel = function() {
  this._showCarousel = true;
  this._updateRepeatedly = true;
  this._showTime = new Date();
};

Carousel.prototype.isCarouselShown = function() {
  return this._showCarousel;
};

Carousel.prototype.hideCarousel = function() {
  // console.log(new Error("Hiding carousel"));
  this._selectedCarouselPlot = null;
  this._selectedCarouselPlotIndex = null;
  this._showCarousel = false;
  this._hideTime = new Date();
  this._viewport.scheduleRepaint();
};

Carousel.prototype.addToCarousel = function(
    node,
    callback,
    thisArg,
    nodeData,
) {
  this._carouselCallbacks.push([callback, thisArg, nodeData]);
  if (!node) {
    throw new Error('Node must not be null');
  }
  if (!node.localPaintGroup && node.root) {
    // Passed a Caret.
    node = node.root();
  }
  if (!node.localPaintGroup()) {
    node.setPaintGroup(true);
  }
  this._carouselPlots.push([node, NaN, NaN, NaN]);
  // console.log("Added to carousel");
};

Carousel.prototype.clearCarousel = function() {
  // console.log("carousel cleared");
  this._carouselPlots.splice(0, this._carouselPlots.length);
  this._carouselCallbacks.splice(0, this._carouselCallbacks.length);
  this._selectedCarouselPlot = null;
  this._selectedCarouselPlotIndex = null;
};

Carousel.prototype.removeFromCarousel = function(node) {
  if (!node) {
    throw new Error('Node must not be null');
  }
  if (!node.localPaintGroup && node.root) {
    // Passed a Caret.
    node = node.root();
  }
  for (const i in this._carouselPlots) {
    if (this._carouselPlots[i][0] === node) {
      // console.log("removed from carousel");
      const removed = this._carouselPlots.splice(i, 1);
      this._carouselCallbacks.splice(i, 1);
      if (this._selectedCarouselPlot === removed) {
        this._selectedCarouselPlot = null;
        this._selectedCarouselPlotIndex = null;
      }
      return removed;
    }
  }
  return null;
};

Carousel.prototype.updateRepeatedly = function() {
  return this._updateRepeatedly;
};

Carousel.prototype.clickCarousel = function(x, y, asDown) {
  if (!this.isCarouselShown()) {
    return false;
  }

  if (this._showTime) {
    const ms = new Date().getTime() - this._showTime.getTime();
    if (ms < CAROUSEL_SHOW_DURATION) {
      // Ignore events that occur so early.
      return true;
    }
  }

  const dist = Math.sqrt(
      Math.pow(Math.abs(x - this._carouselCoords[0]), 2) +
      Math.pow(Math.abs(y - this._carouselCoords[1]), 2),
  );
  if (dist < (this._carouselSize * 0.75) / this.camera().scale()) {
    if (asDown) {
      // console.log("Down events within the inner' +
      //   ' region are treated as 'cancel.'");
      this.hideCarousel();
      this.scheduleCarouselRepaint();
      return true;
    }

    // console.log("Up events within the inner region are ignored.");
    return false;
  } else if (dist > (this._carouselSize * 4) / this.camera().scale()) {
    this.hideCarousel();
    this.scheduleCarouselRepaint();
    // console.log("Click occurred so far outside that' +
    //   ' it is considered its own event.");
    return false;
  }

  const angleSpan = (2 * Math.PI) / this._carouselPlots.length;
  let mouseAngle = Math.atan2(
      y - this._carouselCoords[1],
      x - this._carouselCoords[0],
  );
  // console.log(
  //   alpha_ToDegrees(mouseAngle) +
  //   " degrees = caret " +
  //   i +
  //   " angleSpan = " +
  //   angleSpan);
  if (this._carouselPlots.length == 1 && Math.abs(mouseAngle) > Math.PI / 2) {
    this.hideCarousel();
    this.scheduleCarouselRepaint();
    // console.log("Click occurred so far outside that' +
    //   ' it is considered its own event.");
    return false;
  }
  mouseAngle += Math.PI;
  const i = Math.floor(mouseAngle / angleSpan);

  // Click was within a carousel caret; invoke the listener.
  this.hideCarousel();
  try {
    const callback = this._carouselCallbacks[i][0];
    const thisArg = this._carouselCallbacks[i][1];
    const nodeData = this._carouselCallbacks[i][2];
    callback.call(thisArg, nodeData);
  } catch (ex) {
    console.log('Error occurred while running command:', ex);
  }

  this.scheduleCarouselRepaint();
  return true;
};

Carousel.prototype.mouseOverCarousel = function(x, y) {
  if (!this.isCarouselShown()) {
    return 0;
  }

  const angleSpan = (2 * Math.PI) / this._carouselPlots.length;
  const mouseAngle =
    Math.PI +
    Math.atan2(y - this._carouselCoords[1], x - this._carouselCoords[0]);
  const dist = Math.sqrt(
      Math.pow(Math.abs(x - this._carouselCoords[0]), 2) +
      Math.pow(Math.abs(y - this._carouselCoords[1]), 2),
  );

  if (
    dist < (this._carouselSize * 4) / this.camera().scale() &&
    dist > (BUD_RADIUS * 4) / this.camera().scale()
  ) {
    if (
      this._carouselPlots.length > 1 ||
      Math.abs(mouseAngle - Math.PI) < Math.PI / 2
    ) {
      const i = Math.floor(mouseAngle / angleSpan);
      // console.log(
      //   alpha_ToDegrees(mouseAngle-Math.PI) +
      //   " degrees = caret " +
      //   i +
      //   " angleSpan = " +
      //   angleSpan);
      const selectionAngle = angleSpan / 2 + i * angleSpan - Math.PI;
      if (i != this._selectedCarouselPlotIndex) {
        this._selectedCarouselPlotIndex = i;
        this._selectedCarouselPlot = this._carouselPlots[i];
      }
      if (this._fanPainter) {
        this._fanPainter.setSelectionAngle(selectionAngle);
        this._fanPainter.setSelectionSize(angleSpan);
      }
      this.scheduleCarouselRepaint();
      return 2;
    }
  }
  if (this._fanPainter) {
    this._fanPainter.setSelectionAngle(null);
    this._fanPainter.setSelectionSize(null);
    this._selectedCarouselPlot = null;
    this._selectedCarouselPlotIndex = null;
    this.scheduleCarouselRepaint();
    return 0;
  }
};

Carousel.prototype.showScale = function() {
  return this._showScale;
};

Carousel.prototype.arrangeCarousel = function() {
  if (this._carouselPlots.length === 0) {
    return;
  }

  const angleSpan = (2 * Math.PI) / this._carouselPlots.length;

  const MAX_CAROUSEL_SIZE = 150;

  const now = new Date();
  // Milliseconds
  const showDuration = CAROUSEL_SHOW_DURATION;
  if (this._showTime) {
    let ms = now.getTime() - this._showTime.getTime();
    if (ms < showDuration) {
      ms /= showDuration / 2;
      if (ms < 1) {
        this._showScale = 0.5 * ms * ms;
      } else {
        ms--;
        this._showScale = -0.5 * (ms * (ms - 2) - 1);
      }
    } else {
      this._showScale = 1;
      this._showTime = null;
      this._updateRepeatedly = false;
    }
  }
  // console.log("Show scale is " + this._showScale);

  let minScale = 1;
  this._carouselPlots.forEach(function(carouselData, i) {
    const root = carouselData[0];
    root.commitLayoutIteratively();

    // Set the origin.
    const caretRad =
      Math.PI +
      angleSpan / 2 +
      (i / this._carouselPlots.length) * (2 * Math.PI);
    carouselData[1] =
      2 * this._carouselSize * this._showScale * Math.cos(caretRad);
    carouselData[2] =
      2 * this._carouselSize * this._showScale * Math.sin(caretRad);

    // Set the scale.
    const commandSize = root.extentSize();
    const xMax = MAX_CAROUSEL_SIZE;
    const yMax = MAX_CAROUSEL_SIZE;
    let xShrinkFactor = 1;
    let yShrinkFactor = 1;
    if (commandSize.width() > xMax) {
      xShrinkFactor = commandSize.width() / xMax;
    }
    if (commandSize.height() > yMax) {
      yShrinkFactor = commandSize.height() / yMax;
    }
    // console.log(
    //   commandSize.width(),
    //   commandSize.height(),
    //   1/Math.max(xShrinkFactor,
    //   yShrinkFactor));
    minScale = Math.min(
        minScale,
        this._showScale / Math.max(xShrinkFactor, yShrinkFactor),
    );
  }, this);

  this._carouselPlots.forEach(function(carouselData, i) {
    if (i === this._selectedCarouselPlotIndex) {
      carouselData[3] = 1.25 * minScale;
    } else {
      carouselData[3] = minScale;
    }
  }, this);
};

Carousel.prototype.setOnScheduleRepaint = function(func, thisArg) {
  thisArg = thisArg || this;
  this.onScheduleRepaint = func;
  this.onScheduleRepaintThisArg = thisArg;
};

Carousel.prototype.scheduleCarouselRepaint = function() {
  // console.log("Scheduling carousel repaint.");
  this._carouselPaintingDirty = true;
  if (this.onScheduleRepaint) {
    this.onScheduleRepaint.call(this.onScheduleRepaintThisArg);
  }
};

Carousel.prototype.contextChanged = function(isLost) {
  this._carouselPaintingDirty = true;
  if (this._fanPainter) {
    this._fanPainter.contextChanged(isLost);
  }
  for (const i in this._carouselPlots) {
    if (Object.prototype.hasOwnProperty.call(this._carouselPlots, i)) {
      const carouselData = this._carouselPlots[i];
      const root = carouselData[0];
      root.contextChanged(isLost);
    }
  }
};

Carousel.prototype.paint = function() {
  if (
    !this._updateRepeatedly &&
    (!this._carouselPaintingDirty || !this._showCarousel)
  ) {
    return false;
  }

  // Paint the carousel.
  // console.log("Painting the carousel");
  this.arrangeCarousel();
  for (const i in this._carouselPlots) {
    if (Object.prototype.hasOwnProperty.call(this._carouselPlots, i)) {
      const paintCompleted = this._carouselPlots[i][0].paint(this.window());
    }
  }

  // Paint the background highlighting fan.
  if (!this._fanPainter) {
    this._fanPainter = new FanPainter(this.window());
  } else {
    this._fanPainter.clear();
  }
  const fanPadding = 1.2;
  this._fanPainter.setAscendingRadius(
      this.showScale() * fanPadding * this._carouselSize,
  );
  this._fanPainter.setDescendingRadius(
      this.showScale() * fanPadding * 2 * this._carouselSize,
  );
  this._fanPainter.selectRad(
      0,
      0,
      0,
      Math.PI * 2,
      createColor(1, 1, 1, 1),
      createColor(0.5, 0.5, 0.5, 0.4),
  );

  this._carouselPaintingDirty = false;
  return this._updateRepeatedly;
};

Carousel.prototype.render = function(world) {
  if (!this._showCarousel) {
    return;
  }
  if (this._updateRepeatedly || this._carouselPaintingDirty) {
    this.paint();
  }

  world = matrixMultiply3x3(
      makeScale3x3(1 / this.camera().scale()),
      makeTranslation3x3(this._carouselCoords[0], this._carouselCoords[1]),
      world,
  );

  this._fanPainter.render(world);

  // Render the carousel if requested.
  for (const i in this._carouselPlots) {
    if (Object.prototype.hasOwnProperty.call(this._carouselPlots, i)) {
      const carouselData = this._carouselPlots[i];
      const root = carouselData[0];
      root.renderOffscreen(
          this.window(),
          // scale * trans * world
          matrixMultiply3x3(
              makeScale3x3(carouselData[3]),
              matrixMultiply3x3(
                  makeTranslation3x3(carouselData[1], carouselData[2]),
                  world,
              ),
          ),
          1.0,
      );
    }
  }
};

File: src/graph/Carousel.c
#include "Carousel.h"
#include "Color.h"
#include "../die.h"
#include "../timing.h"
#include "log.h"
#include "initialize.h"
#include <math.h>
#include <stdlib.h>

parsegraph_Carousel* parsegraph_Carousel_new(parsegraph_Camera* camera, float* backgroundColor)
{
    apr_pool_t* pool = 0;
    if(APR_SUCCESS != apr_pool_create(&pool, camera->surface->pool)) {
        parsegraph_die("Failed to create carousel memory pool.");
    }
    parsegraph_Carousel* carousel = apr_palloc(pool, sizeof(*carousel));
    carousel->pool = pool;
    carousel->_updateRepeatedly = 0;
    carousel->_showScale = 1;

    carousel->onScheduleRepaint = 0;
    carousel->onScheduleRepaintThisArg = 0;

    // Carousel-rendered carets.
    carousel->_carouselPaintingDirty = 1;
    carousel->_carouselPlots = parsegraph_ArrayList_new(pool);
    carousel->_carouselCallbacks = parsegraph_ArrayList_new(pool);

    parsegraph_Color_copy(carousel->_backgroundColor, backgroundColor);
    carousel->_camera = camera;

    carousel->_showTime.tv_sec = 0;
    carousel->_hideTime.tv_sec = 0;

    // Location of the carousel, in world coordinates.
    carousel->_carouselX = 0;
    carousel->_carouselY = 0;
    carousel->_carouselSize = 50;

    carousel->_showCarousel = 0;
    carousel->_selectedCarouselPlot = 0;
    carousel->_selectedCarouselPlotIndex = 0;

    carousel->_glyphAtlas = 0;
    carousel->_shaders = 0;

    // GL painters are not created until needed.
    carousel->_fanPainter = 0;

    carousel->_selectedPlot = 0;
    return carousel;
}

void parsegraph_Carousel_destroy(parsegraph_Carousel* carousel)
{
    parsegraph_Carousel_clearCarousel(carousel);
    if(carousel->_fanPainter) {
        parsegraph_FanPainter_destroy(carousel->_fanPainter);
    }
    parsegraph_ArrayList_destroy(carousel->_carouselPlots);
    parsegraph_ArrayList_destroy(carousel->_carouselCallbacks);
    apr_pool_destroy(carousel->pool);
}

parsegraph_Camera* parsegraph_Carousel_camera(parsegraph_Carousel* carousel)
{
    return carousel->_camera;
};

int parsegraph_Carousel_needsRepaint(parsegraph_Carousel* carousel)
{
    return carousel->_carouselPaintingDirty;
};

void parsegraph_Carousel_prepare(parsegraph_Carousel* carousel, parsegraph_GlyphAtlas* glyphAtlas, apr_hash_t* shaders)
{
    carousel->_glyphAtlas = glyphAtlas;
    carousel->_shaders = shaders;
}

void parsegraph_Carousel_moveCarousel(parsegraph_Carousel* carousel, float worldX, float worldY)
{
    carousel->_carouselX = worldX;
    carousel->_carouselY = worldY;
}

void parsegraph_Carousel_setCarouselSize(parsegraph_Carousel* carousel, float size)
{
    carousel->_carouselSize = size;
}

void parsegraph_Carousel_showCarousel(parsegraph_Carousel* carousel)
{
    carousel->_showCarousel = 1;
    carousel->_updateRepeatedly = 1;
    clock_gettime(CLOCK_REALTIME, &carousel->_showTime);
};

int parsegraph_Carousel_isCarouselShown(parsegraph_Carousel* carousel)
{
    return carousel->_showCarousel;
};

int parsegraph_Carousel_isShown(parsegraph_Carousel* carousel)
{
    return parsegraph_Carousel_isCarouselShown(carousel);
}

void parsegraph_Carousel_hideCarousel(parsegraph_Carousel* carousel)
{
    carousel->_selectedCarouselPlot = 0;
    carousel->_selectedCarouselPlotIndex = -1;
    carousel->_showCarousel = 0;
    clock_gettime(CLOCK_REALTIME, &carousel->_hideTime);
};

void parsegraph_Carousel_addToCarousel(parsegraph_Carousel* carousel, parsegraph_Node* node, void(*callback)(void*), void* thisArg)
{
    if(!node) {
        parsegraph_die("Node must not be null");
    }
    struct parsegraph_CarouselCallback* cb = apr_palloc(carousel->pool, sizeof(*cb));
    cb->callback = callback;
    cb->thisArg = thisArg;
    parsegraph_ArrayList_push(carousel->_carouselCallbacks, cb);

    if(!parsegraph_Node_localPaintGroup(node)) {
        parsegraph_Node_setPaintGroup(node, 1);
    }

    parsegraph_CarouselPlot* carouselPlot = malloc(sizeof(parsegraph_CarouselPlot));
    carouselPlot->node = node;
    carouselPlot->x = 0;
    carouselPlot->y = 0;
    carouselPlot->scale = 0;
    parsegraph_ArrayList_push(carousel->_carouselPlots, carouselPlot);
    //parsegraph_log("Added to carousel");
}

void parsegraph_Carousel_clearCarousel(parsegraph_Carousel* carousel)
{
    //console.log("carousel cleared");
    for(int i = 0; i < parsegraph_ArrayList_length(carousel->_carouselPlots); ++i) {
        parsegraph_CarouselPlot* carouselPlot = parsegraph_ArrayList_at(carousel->_carouselPlots, i);
        free(carouselPlot);
    }
    parsegraph_ArrayList_clear(carousel->_carouselPlots);
    parsegraph_ArrayList_clear(carousel->_carouselCallbacks);
    carousel->_selectedCarouselPlot = 0;
    carousel->_selectedCarouselPlotIndex = 0;
}

int parsegraph_Carousel_removeFromCarousel(parsegraph_Carousel* carousel, parsegraph_Node* node)
{
    if(!node) {
        parsegraph_die("Node must not be null");
    }

    for(int i = 0; i < parsegraph_ArrayList_length(carousel->_carouselPlots); ++i) {
        parsegraph_CarouselPlot* carouselPlot = parsegraph_ArrayList_at(carousel->_carouselPlots, i);
        if(carouselPlot->node == node) {
            parsegraph_ArrayList_splice(carousel->_carouselPlots, i, 1);
            if(carousel->_selectedCarouselPlot == carouselPlot) {
                carousel->_selectedCarouselPlot = 0;
                carousel->_selectedCarouselPlotIndex = -1;
            }
            free(carouselPlot);
            return 1;
        }
    }

    return 0;
}

int parsegraph_Carousel_updateRepeatedly(parsegraph_Carousel* carousel)
{
    return carousel->_updateRepeatedly;
}

static float absf(float x)
{
    return x < 0 ? -x : x;
}

int parsegraph_Carousel_clickCarousel(parsegraph_Carousel* carousel, float x, float y, int asDown)
{
    if(!parsegraph_Carousel_isCarouselShown(carousel)) {
        return 0;
    }

    if(carousel->_showTime.tv_sec > 0) {
        struct timespec ts;
        clock_gettime(CLOCK_REALTIME, &ts);
        if(parsegraph_timediffMs(&ts, &carousel->_showTime) < parsegraph_CAROUSEL_SHOW_DURATION) {
            // Ignore events that occur so early.
            return 1;
        }
    }

    // Transform client coords to world coords.
    float* mouseInWorld = matrixTransform2D(
        carousel->pool,
        makeInverse3x3(carousel->pool, parsegraph_Camera_worldMatrix(parsegraph_Carousel_camera(carousel), carousel->pool)),
        x, y
    );
    x = mouseInWorld[0];
    y = mouseInWorld[1];

    float dist = sqrtf(
        powf(absf(x - carousel->_carouselX), 2) +
        powf(absf(y - carousel->_carouselY), 2)
    );
    if(dist < carousel->_carouselSize * .75/parsegraph_Camera_scale(carousel->_camera)) {
        if(asDown) {
            parsegraph_log("Hiding carousel\n");
            parsegraph_Carousel_hideCarousel(carousel);
            parsegraph_Carousel_scheduleCarouselRepaint(carousel);
            return 1;
        }

        //console.log("Up events within the inner region are ignored.");
        return 0;
    }
    else if(dist > carousel->_carouselSize * 4/parsegraph_Camera_scale(carousel->_camera)) {
        parsegraph_Carousel_hideCarousel(carousel);
        parsegraph_Carousel_scheduleCarouselRepaint(carousel);
        //console.log("Click occurred so far outside that it is considered its own event.");
        return 1;
    }

    float angleSpan = 2 * 3.14159 / parsegraph_ArrayList_length(carousel->_carouselPlots);
    float mouseAngle = 3.14159 + atan2f(y - carousel->_carouselY, x - carousel->_carouselX);
    float i = floorf(mouseAngle / angleSpan);

    // Click was within a carousel caret; invoke the listener.
    //console.log(alpha_ToDegrees(mouseAngle) + " degrees = caret " + i);
    parsegraph_Carousel_hideCarousel(carousel);

    struct parsegraph_CarouselCallback* cb = parsegraph_ArrayList_at(carousel->_carouselCallbacks, i);
    cb->callback(cb->thisArg);

    parsegraph_Carousel_scheduleCarouselRepaint(carousel);
    return 1;
};

int parsegraph_Carousel_mouseOverCarousel(parsegraph_Carousel* carousel, float x, float y)
{
    if(!parsegraph_Carousel_isCarouselShown(carousel)) {
        return 0;
    }

    apr_pool_t* pool = 0;
    if(APR_SUCCESS != apr_pool_create(&pool, carousel->pool)) {
        parsegraph_die("Failed to create Carousel mouseover pool");
    }

    float* mouseInWorld = matrixTransform2D(
        pool,
        makeInverse3x3(pool,
            parsegraph_Camera_worldMatrix(parsegraph_Carousel_camera(carousel), pool)
        ),
        x, y
    );
    x = mouseInWorld[0];
    y = mouseInWorld[1];

    float angleSpan = 2 * 3.14159 / parsegraph_ArrayList_length(carousel->_carouselPlots);
    float mouseAngle = 3.14159 + atan2f(y - carousel->_carouselY, x - carousel->_carouselX);
    float dist = sqrtf(
        powf(absf(x - carousel->_carouselX), 2) +
        powf(absf(y - carousel->_carouselY), 2)
    );

    float camScale = parsegraph_Camera_scale(carousel->_camera);
    if(dist < carousel->_carouselSize*4/camScale && dist > parsegraph_BUD_RADIUS*4/camScale) {
        float i = floorf(mouseAngle / angleSpan);
        float selectionAngle = (angleSpan/2 + i * angleSpan) - 3.14159;
        if(i != carousel->_selectedCarouselPlotIndex) {
            carousel->_selectedCarouselPlotIndex = i;
            carousel->_selectedCarouselPlot = parsegraph_ArrayList_at(carousel->_carouselPlots, i);
        }
        if(carousel->_fanPainter) {
            parsegraph_FanPainter_setSelectionAngle(carousel->_fanPainter, selectionAngle);
            parsegraph_FanPainter_setSelectionSize(carousel->_fanPainter, angleSpan);
        }
        parsegraph_Carousel_scheduleCarouselRepaint(carousel);
        return 2;
    }
    else if(carousel->_fanPainter) {
        parsegraph_FanPainter_setSelectionAngle(carousel->_fanPainter, 0);
        parsegraph_FanPainter_setSelectionSize(carousel->_fanPainter, 0);
        carousel->_selectedCarouselPlotIndex = -1;
        carousel->_selectedCarouselPlot = 0;
        parsegraph_Carousel_scheduleCarouselRepaint(carousel);
    }
    return 0;
}

float parsegraph_Carousel_showScale(parsegraph_Carousel* carousel)
{
    return carousel->_showScale;
}

void parsegraph_Carousel_arrangeCarousel(parsegraph_Carousel* carousel)
{
    size_t numPlots = parsegraph_ArrayList_length(carousel->_carouselPlots);
    if(numPlots == 0) {
        return;
    }

    float angleSpan = 2.0*3.14159/((float)numPlots);

    float parsegraph_MAX_CAROUSEL_SIZE = 150.0f;

    struct timespec now;
    clock_gettime(CLOCK_REALTIME, &now);
    // Milliseconds
    float showDuration = parsegraph_CAROUSEL_SHOW_DURATION;
    if(carousel->_showTime.tv_sec > 0) {
        float ms = parsegraph_timediffMs(&carousel->_showTime, &now);
        if(ms < showDuration) {
            ms /= showDuration/2.0f;
            if(ms < 1.0f) {
                carousel->_showScale = 0.5f*ms*ms;
            }
            else {
                ms--;
                carousel->_showScale = -0.5f*(ms*(ms-2.0f)-1.0f);
            }
        }
        else {
            carousel->_showScale = 1;
            carousel->_showTime.tv_sec = 0;
            carousel->_updateRepeatedly = 0;
        }
        parsegraph_log("Carousel shown for %fms has show scale of %f\n", ms, carousel->_showScale);
    }

    float minScale = 1;
    for(int i = 0; i < parsegraph_ArrayList_length(carousel->_carouselPlots); ++i) {
        parsegraph_CarouselPlot* carouselData = parsegraph_ArrayList_at(carousel->_carouselPlots, i);
        parsegraph_Node* root = carouselData->node;
        parsegraph_Node_commitLayoutIteratively(root, 0);

        // Set the origin.
        float caretRad = 3.14159 + angleSpan/2.0f + ((float)i / (float)numPlots) * (2.0f * 3.14159);
        carouselData->x = 2.0f*carousel->_carouselSize * carousel->_showScale * cosf(caretRad);
        carouselData->y = 2.0f*carousel->_carouselSize * carousel->_showScale * sinf(caretRad);
        parsegraph_log("Placing carousel node %d's root at (%f, %f)\n", i, carouselData->x, carouselData->y);

        // Set the scale.
        float extentSize[2];
        parsegraph_Node_extentSize(root, extentSize);
        float xMax = parsegraph_MAX_CAROUSEL_SIZE;
        float yMax = parsegraph_MAX_CAROUSEL_SIZE;
        float xShrinkFactor = 1.0f;
        float yShrinkFactor = 1.0f;
        if(extentSize[0] > xMax) {
            xShrinkFactor = extentSize[0] / xMax;
        }
        if(extentSize[1] > yMax) {
            yShrinkFactor = extentSize[1] / yMax;
        }
        float maxShrinkFactor = (xShrinkFactor > yShrinkFactor) ? xShrinkFactor : yShrinkFactor;
        //fprintf(stderr, "%f %f %f\n", extentWidth, extentHeight, 1/maxShrinkFactor);
        float thisScale = carousel->_showScale/maxShrinkFactor;
        if(thisScale < minScale) {
            minScale = thisScale;
        }
    }

    for(int i = 0; i < parsegraph_ArrayList_length(carousel->_carouselPlots); ++i) {
        parsegraph_CarouselPlot* carouselData = parsegraph_ArrayList_at(carousel->_carouselPlots, i);
        if(i == carousel->_selectedCarouselPlotIndex) {
            carouselData->scale = 1.25f*minScale;
        }
        else {
            carouselData->scale = minScale;
        }
    }
}

void parsegraph_Carousel_setOnScheduleRepaint(parsegraph_Carousel* carousel, void(*func)(void*), void* thisArg)
{
    if(!thisArg) {
        thisArg = carousel;
    }
    carousel->onScheduleRepaint = func;
    carousel->onScheduleRepaintThisArg = thisArg;
}

void parsegraph_Carousel_scheduleCarouselRepaint(parsegraph_Carousel* carousel)
{
    parsegraph_logEntercf("Repaint scheduling", "Scheduling carousel repaint.\n");
    carousel->_carouselPaintingDirty = 1;
    if(carousel->onScheduleRepaint) {
        carousel->onScheduleRepaint(carousel->onScheduleRepaintThisArg);
    }
    parsegraph_logLeave();
}

parsegraph_GlyphAtlas* parsegraph_Carousel_glyphAtlas(parsegraph_Carousel* carousel)
{
    return carousel->_glyphAtlas;
}

float* parsegraph_Carousel_backgroundColor(parsegraph_Carousel* carousel)
{
    return carousel->_backgroundColor;
}

void parsegraph_Carousel_setBackgroundColor(parsegraph_Carousel* carousel, float* backgroundColor)
{
    parsegraph_Color_copy(carousel->_backgroundColor, backgroundColor);
    parsegraph_Carousel_scheduleCarouselRepaint(carousel);
}

void parsegraph_Carousel_paint(parsegraph_Carousel* carousel)
{
    if(!carousel->_updateRepeatedly && (!carousel->_carouselPaintingDirty || !carousel->_showCarousel)) {
        return;
    }

    // Paint the carousel.
    parsegraph_logEntercf("Carousel paints", "Painting the carousel");
    parsegraph_Carousel_arrangeCarousel(carousel);
    for(int i = 0; i < parsegraph_ArrayList_length(carousel->_carouselPlots); ++i) {
        parsegraph_CarouselPlot* carouselData = parsegraph_ArrayList_at(carousel->_carouselPlots, i);
        parsegraph_Node* paintGroup = carouselData->node;
        parsegraph_PAINTING_GLYPH_ATLAS = parsegraph_Carousel_glyphAtlas(carousel);
        parsegraph_Node_paint(
            paintGroup,
            parsegraph_Carousel_backgroundColor(carousel),
            parsegraph_Carousel_glyphAtlas(carousel),
            carousel->_shaders,
            0
        );
        parsegraph_PAINTING_GLYPH_ATLAS = 0;
    }

    // Paint the background highlighting fan.
    if(!carousel->_fanPainter) {
        carousel->_fanPainter = parsegraph_FanPainter_new(carousel->pool, carousel->_shaders);
    }
    else {
        parsegraph_FanPainter_clear(carousel->_fanPainter);
    }
    float fanPadding = 1.2;
    parsegraph_FanPainter* fanPainter = carousel->_fanPainter;
    parsegraph_log("Carousel show scale currently %f\n", parsegraph_Carousel_showScale(carousel));
    float ascRadius = parsegraph_Carousel_showScale(carousel) * fanPadding * carousel->_carouselSize;
    parsegraph_log("Ascending radius is %f\n", ascRadius);
    parsegraph_FanPainter_setAscendingRadius(fanPainter, ascRadius);
    float descRadius = parsegraph_Carousel_showScale(carousel) * fanPadding * 2 * carousel->_carouselSize;
    parsegraph_FanPainter_setDescendingRadius(fanPainter, descRadius);
    parsegraph_log("Descending radius is %f\n", descRadius);

    float startColor[] = {1, 1, 1, 1};
    float endColor[] = {.5, .5, .5, .4};
    parsegraph_FanPainter_selectRad(
        fanPainter,
        0, 0,
        0, 3.14159 * 2.0f,
        startColor, endColor
    );

    parsegraph_log("Carousel show scale is %f\n", carousel->_showScale);
    carousel->_carouselPaintingDirty = 0;
    parsegraph_logLeave();
}

void parsegraph_Carousel_render(parsegraph_Carousel* carousel, float* world)
{
    if(!carousel->_showCarousel) {
        return;
    }
    if(carousel->_updateRepeatedly || carousel->_carouselPaintingDirty) {
        parsegraph_Carousel_paint(carousel);
    }

    float dest[9];
    float trans[9];
    // translation * scale * world
    makeScale3x3I(dest, 1.0f/parsegraph_Camera_scale(carousel->_camera), 1.0f/parsegraph_Camera_scale(carousel->_camera));
    makeTranslation3x3I(trans, carousel->_carouselX, carousel->_carouselY);
    matrixMultiply3x3I(dest, dest, trans);
    matrixMultiply3x3I(dest, dest, world);
    world = dest;

    parsegraph_FanPainter_render(carousel->_fanPainter, world);

    // Render the carousel if requested.
    for(int i = 0; i < parsegraph_ArrayList_length(carousel->_carouselPlots); ++i) {
        parsegraph_CarouselPlot* carouselPlot = parsegraph_ArrayList_at(carousel->_carouselPlots, i);
        parsegraph_Node* root = carouselPlot->node;

        makeTranslation3x3I(trans, carouselPlot->x, carouselPlot->y);
        parsegraph_log("Carousel plot at (%f, %f)\n", carouselPlot->x, carouselPlot->y);
        float sc[9];
        makeScale3x3I(sc, carouselPlot->scale, carouselPlot->scale);
        float carouselNodeMat[9];
        matrixMultiply3x3I(carouselNodeMat, sc, trans);
        matrixMultiply3x3I(carouselNodeMat, carouselNodeMat, world);

        parsegraph_Node_renderIteratively(root, carouselNodeMat, 0);
    }
}
